typings/externs.d.ts

```diff
@@ -4,139 +4,141 @@
-  displayValue: string;
-  debugString?: string;
-  score: number;
-  scoreDisplayMode: string;
-  error?: boolean;
-  description: string;
-  name: string;
-  helpText?: string;
-  extendedInfo?: {value: string};
-}
-
-export interface AuditFullResults {
-  [metric: string]: AuditFullResult;
-}
-
-export interface Results {
-  url: string;
-  audits: AuditFullResults;
-  lighthouseVersion: string;
-  artifacts?: Object;
-  initialUrl: string;
-  generatedTime: string;
-}
-
-export interface LaunchedChrome {
-  pid: number;
-  port: number;
-  kill: () => Promise<{}>;
-}
-
-export interface LighthouseError extends Error {
-  code?: string;
-  friendlyMessage?: string;
-}
-
-export interface TraceEvent {
-  name: string;
-  args: any;
-  tid: number;
-  ts: number;
-  dur: number;
-}
-
-export interface NetworkRequest {
-  requestId: string;
-  connectionId: string;
-  connectionReused: boolean;
-
-  url: string;
-  protocol: string;
-  origin: string | null;
-  parsedURL: DevToolsParsedURL;
-
-  startTime: number;
-  endTime: number;
-
-  transferSize: number;
-
-  _initiator: NetworkRequestInitiator;
-  _timing: NetworkRequestTiming;
-  _resourceType: any;
-  priority(): 'VeryHigh' | 'High' | 'Medium' | 'Low';
-}
-
-export interface NetworkRequestInitiator {
-  type: 'script' | 'parser';
-}
-
-export interface NetworkRequestTiming {
-  connectStart: number;
-  connectEnd: number
-  sslStart: number;
-  sslEnd: number;
-  sendStart: number;
-  sendEnd: number;
-  receiveHeadersEnd: number;
-}
-
-export interface DevToolsParsedURL {
-  scheme: string;
-  host: string;
-}
-
-export interface DevToolsJsonTarget {
-  description: string;
-  devtoolsFrontendUrl: string;
-  id: string;
-  title: string;
-  type: string;
-  url: string;
-  webSocketDebuggerUrl: string;
+declare namespace LH {
+
+  export interface Flags {
+    _: string[];
+    port: number;
+    chromeFlags: string;
+    output: any;
+    outputPath: string;
+    saveAssets: boolean;
+    view: boolean;
+    maxWaitForLoad: number;
+    logLevel: string;
+    hostname: string;
+    blockedUrlPatterns: string[];
+    extraHeaders: string;
+    enableErrorReporting: boolean;
+    listAllAudits: boolean;
+    listTraceCategories: boolean;
+    auditMode: boolean;
+    gatherMode: boolean;
+    configPath?: string;
+    perf: boolean;
+    mixedContent: boolean;
+    verbose: boolean;
+    quiet: boolean;
+  }
+
+  export interface Config {}
+
+  export interface AuditResult {
+    rawValue: boolean | number | null;
+    displayValue?: string;
+    debugString?: string;
+    score?: number;
+    extendedInfo?: {value: string};
+    notApplicable?: boolean;
+    error?: boolean;
+    // TODO: define details
+    details?: object;
+  }
+
+  export interface AuditResults {
+    [metric: string]: AuditResult;
+  }
+
+  export interface AuditFullResult extends AuditResult {
+    displayValue: string;
+    score: number;
+    scoreDisplayMode: 'numeric' | 'binary';
```

Comment:
Audit.ScoringModes?

Reply:
`Audit.ScoringModes` is scoped to `audit.js`, so the externs would need to import `Audit` to have access. I think that's in our future anyways, but not worth it for this (it changes other things when our `.d.ts` file imports things because then it too becomes a module...)

Reply:
The future plan is to much as much as external facing stuff as possible to a .d.ts file anyway right?

Reply:
I'm not sure. There's also what we want global vs scoped. What we produce for others to link against our public API may or may not be tied to that :)

Reply:
I think I was also wrong here. All the typedefs in `Audit` appear to be entirely local to the file and aren't exported, even though they're on an object that was, so all of these may need to be in d.ts files unless only used inside this file (that includes subclasses, so right now no audits can see the typedef of `Audt.Meta` even though they all import `Audit`